datalad_dataverse/remote.py

```diff
@@ -92,111 +136,311 @@ def api(self):
+
+        return self._files_all
+
+    @property
+    def files_latest(self):
+        """Cache of files in the latest version of the dataverse dataset.
+
+        This refers to the DRAFT version (if there is any) or the latest
+        published version otherwise. That's the version pushes go into. Hence,
+        this is needed to determine whether we need to replace a file on
+        upload, while the complete list in `self.files_all` is relevant for key
+        retrieval (which can come from earlier dataverse versions of the
+        dataset).
+        """
+
+        if self._files_latest is None:
+            # Latest version in self.dataset is first entry.
+            self._files_latest = {
+                f['dataFile']['id']: Path(f.get('directoryLabel', '')) / f['dataFile']['filename']
+                for f in self.dataset[0]['files']}
+
+        return self._files_latest
+
+    def remove_from_filelist(self, id):
+        """Update self.filelist after removal"""
+        if self._files_all is None:
+            self.message(f"FAILED Removing {id} from file list failed. "
+                         f"'_filelist' does not exist.",
+                         type='debug')
+            return  # not yet a reason to crash
+        self._files_all.pop(id)
+        try:
+            self._files_latest.pop(id)
+        except KeyError:
+            # This is alright, we removed from _all, it may not have been in
+            # latest, just calling here to be sure we are consistent between the
+            # two.
+            pass
+
+    def add_to_filelist(self, d):
+        """Update self.filelist after upload
+
+        d: dict
+          dataverse description dict of the file; this dict is in the list
+          'data.files' of the response to a successful upload
+        """
+        entry = (d['dataFile']['id'],
+                 Path(d.get('directoryLabel', '')) /
+                    d['dataFile']['filename']
+                 )
+        self._files_all[entry[0]] = entry[1]
+        self._files_latest[entry[0]] = entry[1]
+
     def prepare(self):
         # trigger API instance in order to get possibly auth/connection errors
         # right away
         self.api
+        # TEMP: get it from the start for now. Needs RF, though. Anything
+        # --fast/--force, may not require it at all.
+        self.dataset
+        self.files_all
+        self.files_latest
 
     def checkpresent(self, key):
-        dataset = self.api.get_dataset(identifier=self.doi)
-
-        datafiles = dataset.json()['data']['latestVersion']['files']
-        if next((item for item in datafiles if item['dataFile']['filename'] == key), None):
-            return True
+        known_state = self.annex.getstate(key)
+        if known_state:  # empty string otherwise
+            return int(known_state) in self.files_all.keys()
         else:
-            return False
+            return Path(key) in [p for p in self.files_all.values()]
```

Comment:
Not confident I got what is happening here. When annex knows the file, we check whether we ID is in the remote dump -- seems kosher. But when annex does not have a record of the file, we go with anything that has the same path/filename? That is probably not a good thing.